### ISSUE TITLE
Passing incompatible arguments to usort should result in error

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#1109]
https://getpsalm.org/r/b20e30baf2
```php
<?php
  $arr = [[1], [2], [3]];
  usort($arr, function (string $a, string $b): int { return strcmp($a, $b); });
```
Psalm sees no problems here, but this code will result in fatal error. 

If Psalm could check argument paradoxes (#742) and supported in/out params (#1088) `usort` could be simply declared as:
```php
 /**
   * @template T
   * @param array<mixed,T> $arr
   * @param-out array<int,T> $arr
   * @param callable(T,T):int $callback
   */
  function _usort(array &$arr, callable $callback): bool {}
```

**Comment by muglug** [comment]
Turns out the fix for in #742 wasn't sufficient to catch this bug.

To catch this in the general case, there would need to be two sweeps through the function arguments when templated types are used - the first sweep would just gather up template constraints - in this case, `T >= int`, `T <= string`. Then there'd be an attempt to resolve them to a single type (which would result in the failure).

**Comment by weirdan** [comment]
:+1:
